network/v1alpha1/types_dnsnameresolver.go

```diff
@@ -0,0 +1,129 @@
+	// '*' can be used at the beginning of the wildcard DNS name. For example, '*.example.com.'
+	// will match 'sub1.example.com.' but won't match 'sub2.sub1.example.com.'
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9]([-A-Za-z0-9]*[A-Za-z0-9])?\.)+[A-Za-z0-9]([-A-Za-z0-9]*[A-Za-z0-9])?\.$
+	// +kubebuilder:validation:MaxLength=254
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="spec.name is immutable"
+	Name string `json:"name"`
+}
+
+// DNSNameResolverStatus defines the observed status of DNSNameResolver.
+type DNSNameResolverStatus struct {
+	// resolvedNames contains a list of matching DNS names and their corresponding IP addresses
+	// along with their TTL and last DNS lookup times.
+	// +listType=map
+	// +listMapKey=dnsName
+	// +patchMergeKey=dnsName
+	// +patchStrategy=merge
+	// +optional
+	ResolvedNames []DNSNameResolverResolvedName `json:"resolvedNames,omitempty" patchStrategy:"merge" patchMergeKey:"dnsName"`
+}
+
+// DNSNameResolverResolvedName describes the details of a resolved DNS name.
+type DNSNameResolverResolvedName struct {
+	// conditions provide information about the state of the DNS name.
+	// Known .status.conditions.type is: "Degraded"
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+	// dnsName is the resolved DNS name matching the name field of DNSNameResolverSpec. This field can
+	// store both regular and wildcard DNS names which match the spec.name field. When the spec.name
+	// field contains a regular DNS name, this field will store the same regular DNS name after it is
+	// successfully resolved. When the spec.name field contains a wildcard DNS name, each resolvedName.dnsName
+	// will store the regular DNS names which match the wildcard DNS name and have been successfully resolved.
+	// If the wildcard DNS name can also be successfully resolved, then this field will store the wildcard
+	// DNS name as well.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9]([-A-Za-z0-9]*[A-Za-z0-9])?\.)+[A-Za-z0-9]([-A-Za-z0-9]*[A-Za-z0-9])?\.$
+	// +kubebuilder:validation:MaxLength=254
+	DNSName string `json:"dnsName"`
+	// resolvedAddresses gives the list of associated IP addresses and their corresponding TTLs and last
+	// lookup times for the dnsName.
+	// +kubebuilder:validation:Required
+	// +listType=map
+	// +listMapKey=ip
+	ResolvedAddresses []DNSNameResolverResolvedAddress `json:"resolvedAddresses"`
+	// resolutionFailures keeps the count of how many consecutive times the DNS resolution failed
+	// for the dnsName. If the DNS resolution succeeds then the field will be set to zero. Upon
+	// every failure, the value of the field will be incremented by one. The details about the DNS
+	// name will be removed, if the value of resolutionFailures reaches 5 and the TTL of all the
+	// associated IP addresses have expired.
```

Comment:
i still don't think that counting resolution failures is the best means of resetting DNS name details.  It means something entirely different if `resolutionFailures` increments from 0 to 5 over an hour as opposed to over a year.  Which details about the DNS name are removed on the 5th failure when the TTL has expired?  Why not just remove or refresh when the TTL expires?  Or, if we want to keep the resolution failures, why not mark the status as Stale or Pending, a new condition?

Reply:
Also, this count should not be used for wildcard DNS names.  If there are 12 resolved DNS names from the wildcard name, the resolution failures should be counted per resolved DNS name, not per wildcard.

Reply:
> increments from 0 to 5 over an hour as opposed to over a year.

The resolution failure count is consecutive, so each time a successful lookup is made the counter is reset to 0

> Also, this count should not be used for wildcard DNS names. If there are 12 resolved DNS names from the wildcard name, the resolution failures should be counted per resolved DNS name, not per wildcard.

@arkadeepsen I feel like we may have already discussed this, but I can't remember now, is there a reason the field is on this scope rather than as part of a resolved address?

Reply:
> Which details about the DNS name are removed on the 5th failure when the TTL has expired? Why not just remove or refresh when the TTL expires?

The `resolutionFailures` field will be used to remove all the information of a DNS name when all the IP addresses have expired. Whenever the TTL of an IP address associated to a DNS name expires, a DNS resolution request is sent for the DNS name. Now, if this resolution fails, we should not remove all the IP addresses. In this case, the `resolutionFailures` field will be incremented by 1. The resolution will be retried after an interval of 5 seconds. In the meantime, the IP address whose TTL expired will be removed after a grace period.  If the resolution fails consecutively 5 times and also the TTL of all the IP addresses have expired, then it'll be assumed that the DNS name does not have a DNS entry and the information about the DNS name will be removed. This scenario is more relevant for removing a DNS name from the `resolvedNames` field when the CR is for a wildcard DNS name, but it also applies to a regular DNS name CR.

>  Or, if we want to keep the resolution failures, why not mark the status as Stale or Pending, a new condition?

Whenever there is a resolution failure the status of the CR will be set to Degraded=true. Whenever, the resolution succeeds it will be set to Degraed=false.

> @arkadeepsen I feel like we may have already discussed this, but I can't remember now, is there a reason the field is on this scope rather than as part of a resolved address?

The field is in this scope as the field will be used to remove all the information about the DNS name from the `resolvedNames` list. The `ttlSeconds` field will be used to remove the IP address with a grace period after the TTL has expired.

Reply:
It might help to clarify the godoc for a wildcard vs regular DNS name.  I expect it to be different, because if `resolutionFailures` is on the DNS name scope, and the DNS name is a wildcard, I don't think it's a good idea to remove all the information from `resolvedNames` that are associated with a wildcard DNS name.  Does that clarify my point?

Reply:
The behavior is same for both wildcard and regular DNS names. As we are only removing the information of the DNS name after 5 consecutive resolution failure and when the TTL of all the IP addresses have expired, it means that we can safely remove the information. In the case of regular DNS name, only one entry will be there in the `resolvedNames` list, whereas in the case of wildcard DNS name the list will contain all the DNS names that match the wildcard DNS name. If the wildcard DNS name has an A or AAAA record then the list will contain the wildcard DNS name as well, or else it will contain the regular DNS names which match it. If we are removing the wildcard DNS name's entry from the `resolvedNames` list, then the other regular DNS names will still continue to remain in the `resolvedNames` list for the wildcard DNS name. Here only the individual DNS name information is going to get removed, even in the case of a wildcard DNS name which has an A or AAAA record.

Reply:
For example, if we have a `resolvedName` list for `*.example.com` as the following:
````yaml
resolvedNames:
- dnsName: "*.example.com"
  resolvedAddresses:
    - ip: "1.1.1.1"
      ttlSeconds: 0
  resolutionFailures: 5
- dnsName: "www.example.com"
  resolvedAddresses:
    - ip: "2.2.2.2"
      ttlSeconds: 30
  resolutionFailures: 0 
````
Then the entry for `*.example.com` will be removed as the `resolutionFailures` has reached 5 and the TTL for the associated IP address has expired. The `resolvedNames` will be updated to the following:
````yaml
resolvedNames:
- dnsName: "www.example.com"
  resolvedAddresses:
    - ip: "2.2.2.2"
      ttlSeconds: 30
  resolutionFailures: 0 
````

Reply:
Okay, that makes sense. Except what would be a resolution of that wildcard dnsName?  Can someone actually get a success when trying to resolve *.example.com?

Reply:
If the A or AAAA record exists for the wildcard DNS name, then the response would be a successful lookup fetching the currently associated IP addresses. If the record doesn't exist, then it'll result in a failure.